### ISSUE TITLE
Holmake raises Subscript in examples/formal-languages/regular

### ISSUE THREAD

**Issue by konrad-slind** [HOL-Theorem-Prover/HOL#549]
I've seen Holmake raise Subscript in other directories as well. First time I have seen it in ...regular

guardol@guardol-host:~/Projects/HOL/examples/formal-languages/regular$ Holmake
*** Switching to execute /home/guardol/Projects/HOL/bin/Holmake
*** (As we are in/under its HOLDIR)
]0;Holmake: ~/Projects/HOL/examples/balanced_bst]0;Holmake: ~/Projects/HOL/examples/balanced_bstWorking in $(HOLDIR)/examples/balanced_bst
]0;Holmake: .]0;Holmake: ~/Projects/HOL/examples/formal-languages]0;Holmake: ~/Projects/HOL/examples/formal-languagesWorking in $(HOLDIR)/examples/formal-languages
]0;Holmake: .]0;Holmake: ~/Projects/HOL/examples/formal-languages/context-free]0;Holmake: ~/Projects/HOL/examples/formal-languages/context-freeWorking in $(HOLDIR)/examples/formal-languages/context-free
]0;Holmake: .]0;Holmake: .Working in $(HOLDIR)/examples/formal-languages/regular
Holmake failed with exception: Subscript

**Comment by xrchz** [comment]
Is this deterministic / easily reproducible?

**Comment by konrad-slind** [comment]
Not sure it's deterministic. A fresh clone of HOL builds fine, and
examples/formal-languages/regular then builds fine.
Needs more investigation. Presumably it's something to do with state that's
not being cleaned out by Holmake cleanAll, or freshness of files. In any
case, the Subscript exception shouldn't get raised.

Konrad.

**Comment by mn200** [comment]
Which version of Poly/ML is it with?  (Totally agree it's a painful bug.)

**Comment by konrad-slind** [comment]
Just found the problem. I was doing something stupid (made dotted
identifier X.id occur inside structure X) and that (a) makes an invocation
of Holmake throw Subscript, (b) this is persistent: even after commenting
out the stupidity and doing a Holmake cleanAll followed by Holmake, I still
get the Subscript being thrown, and (c) some state I don't know about is
being written because if I then go to HOLDIR and do a bin/build cleanAll
followed by a bin/build that will also die horribly.

To replicate: add the following line after the definition of pp_sexp in
..../regular/Regexp_Type.sml and invoke Holmake.

val _ = PolyML.addPrettyPrinter (fn d => fn _ => fn r =>
Regexp_Type.pp_regexp r);

**Comment by mn200** [comment]
Ha!  Thanks for the reproducibility!

**Comment by konrad-slind** [comment]
Yes, I moved that line from one of the (later) test files, and forgot to
remove the dotted identifier. Was trying to see if I could install the
printer from inside a module. (My memory said that the install had to be at
the top level, but was going to try anyway.) Then all hell broke loose, I
didn't see the cause for a while. Interesting that the bug seems to nuke
seemingly unrelated stuff.

Konrad.

**Comment by mn200** [comment]
I can't see a `pp_sexp` in the file as checked in but am trying the line you quoted just in the file towards the end...

**Comment by mn200** [comment]
After making a clean build, and then editing `Regexp_Type.sml` to include that line, I don't see anything.

**Comment by mn200** [comment]
````
diff --git a/examples/formal-languages/regular/Regexp_Type.sml b/example
index 5961560..24a7f8e 100644
--- a/examples/formal-languages/regular/Regexp_Type.sml
+++ b/examples/formal-languages/regular/Regexp_Type.sml
@@ -1366,4 +1366,7 @@ fun fromString s = tree_to_regexp (string_to_tree 
 fun fromQuote q = tree_to_regexp (quote_to_tree q)
  handle e => raise wrap_exn "Regexp_Type" "fromQuote" e;
 
+val _ = PolyML.addPrettyPrinter (fn d => fn _ => fn r =>
+Regexp_Type.pp_regexp r);
+
 end
````

**Comment by mn200** [comment]
Take it back; Holmake tries to rebuild `charsetTheory` (which refers to `Regexp_Type`) and this fails with a reasonable error message.

**Comment by mn200** [comment]
Some further testing reveals that there seems to be a difference in behaviour when running under an emacs shell.

**Comment by mn200** [comment]
Closed by 1c9452f
